### PR TITLE
Happy thoughts api Johan Löfgren

### PR DESCRIPTION
**Pull request by johlofg**

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,29 +1,78 @@
...
 import express from 'express'
-import bodyParser from 'body-parser'
 import cors from 'cors'
 import mongoose from 'mongoose'
+import listEndpoints from 'express-list-endpoints'
 
 const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/happyThoughts"
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
+// eslint-disable-next-line max-len
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true, useCreateIndex: true })
```
Comment (Polkapetunian) [inline server.js:9]: I did not know about useCreateIndex but I think it is used to get rid a deprecation warning. Good idea!

File: server.js [hunk 2]
```diff
@@ -1,29 +1,78 @@
...
+const Thought = mongoose.model('Thought', thoughtSchema)
+
 app.use(cors())
-app.use(bodyParser.json())
+app.use(express.json())
 
-// Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
```
Comment (Polkapetunian) [inline server.js:38]: Good that you changed this to display a list of the endpoints!👍

File: server.js [hunk 3]
```diff
@@ -1,29 +1,78 @@
...
+    type: String, 
+    required: true,    
+    minlength: 5,
+    maxlength: 140
+  },   
+  hearts: {
+    type: Number,
+    default: 0
+  },
+  CreatedAt: {
```
Comment (Polkapetunian) [inline server.js:26]: Why not camelCase in createdAt?

File: server.js [hunk 4]
```diff
@@ -1,29 +1,78 @@
...
+app.post('/thoughts', async (req, res) => {
+  try {
+    const newThought = await new Thought({ message: req.body.message }).save()
+    res.json(newThought)
+  } catch (error) {    
+    res.status(400).json({ message: 'Invalid request', error })
+  } 
+})
+
+app.delete('/thoughts/:id', async (req, res) => {
```
Comment (Polkapetunian) [inline server.js:55]: It's fun with a delete functionality but it was not supposed to be there according to the brief I think...

File: server.js [hunk 5]
```diff
@@ -1,29 +1,78 @@
...
+  } catch (error) {
+    res.status(400).json({ message: "Invalid request", error })
+  }
+})
+
+app.post('/thoughts/:id/likes', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const updatedThought = await Thought.findByIdAndUpdate(id, { $inc: { hearts: 1 } })
```
Comment (Polkapetunian) [inline server.js:69]: To show the number of likes after the update use 
Thought.findByIdAndUpdate(  id,  {   $inc: { hearts: 1  }  },  {  new: true  } )

File: server.js [hunk 6]
```diff
@@ -1,29 +1,78 @@
...
+    type: Number,
+    default: 0
+  },
+  CreatedAt: {
+    type: Date,
+    default: Date.now
+  }  
+})
+
+const Thought = mongoose.model('Thought', thoughtSchema)
```
Comment (Polkapetunian) [inline server.js:32]: Nice and clear code!

**Review by Polkapetunian — APPROVED** [review]
I think the blue goals are reached (wonder a bit about the possibility to delete though...). I tested the live fullstack app and also the API with Postman and it works great! The error messages are very good! I would correct the text in the Readme for portfolio purposes. There are some typos and grammar issues.

File: server.js [hunk 7]
```diff
@@ -1,29 +1,78 @@
...
 const port = process.env.PORT || 8080
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
+const thoughtSchema = new mongoose.Schema({
+  message: {
+    type: String, 
+    required: true,    
+    minlength: 5,
+    maxlength: 140
```
Comment (KarolCislowski) [inline server.js:20]: Nicely configured field 👍🏻

File: server.js [hunk 8]
```diff
@@ -1,29 +1,78 @@
...
+app.use(express.json())
 
-// Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+app.get('/thoughts', async (req, res) => {  
+  const allThoughts = await Thought.find().sort({ CreatedAt: -1 }).limit(20)
```
Comment (KarolCislowski) [inline server.js:42]: 👍🏻

File: server.js (hunk 5 above)
Comment (KarolCislowski) [reply server.js:69]: 💯

File: server.js [hunk 9]
```diff
@@ -1,29 +1,78 @@
...
+
+app.delete('/thoughts/:id', async (req, res) => {
+  const { id } = req.params
+  try {    
+    const deletedThought = await Thought.deleteOne({ _id: id })
+    res.json(deletedThought) 
+  } catch (error) {
+    res.status(400).json({ message: "Invalid request", error })
+  }
+})
```
Comment (KarolCislowski) [inline server.js:63]: Remember to restrict access to this endpoint because not all users should have access to this functionality 🤓

**Review by KarolCislowski — APPROVED** [review]
Great API 🥇

**Pull request closed without merging**